docs/proposals/20220215-obsctl.md

```diff
@@ -0,0 +1,67 @@
+# obsctl: A CLI to interact with Observatorium
+
+* **Owners:**
+  * `@onprem` `@saswatamcode`
+
+* **Related Tickets:**
+  * https://issues.redhat.com/browse/MON-2199
+
+* **Other docs:**
+  * None
+
+## Why
+
+We can interact with [Observatorium API](https://github.com/observatorium/api) with a convenient and dedicated CLI, which provides capabilities to persist auth data, and allows us to easily & quickly get up and running with our operations.
+
+## Pitfalls of Current Solution
+
+Currently, we need to manually run [token-refresher](https://github.com/observatorium/token-refresher) with [OIDC](https://openid.net/connect/) information for a particular tenant, and then make cURL requests through that. This is not optimal and generally confusing/trickier for users, both old and new.
+
+## Goals
+
+obsctl will,
+* Manage authentication configuration for multiple tenants by saving them locally
+* Allow users to switch between tenants conveniently
+* Configure and view rules for a tenant
+* View series, rules, and labels for a tenant
+* Later on, support more such one-off operations for other domains as well (logs, traces, alert-routing configs)
+
+## Non-Goals
+
+* Adding and dynamically configuring tenants through obsctl (No such admin API is available within Observatorium, nor is it planned)
+
+## Audience of this proposal
+
+Observatorium Devs and Users.
+
+## How
+
+As obsctl is a CLI, authentication options (OIDC) need to be persisted across sessions. obsctl aims to achieve this by saving those configurations locally at some configuration directory and maintaining a "current" tenant.
+
+The "current" tenant will basically represent the logged-in tenant and all operations will be done based on this "current" tenant configuration. In case, the "current" tenant is empty, i.e, there is no tenant currently logged-in, obsctl will error out.
+
+obsctl will also allow the user to switch from the "current" tenant to another one. This can be achieved by creating a map of tenant configurations that are uniquely referenceable.
+
+### Authentication Commands to be supported by obsctl
+
+* `login <OIDC flags>`: Login as a tenant and save their configuration locally.
+* `logout`: Logout current tenant.
+* `switch <tenant>`: Switch from one tenant to another.
```

Comment:
So the flow will be:

obsctl switch tenant1
obsctl login 

In order to login as tenant? Quite opaque, maybe there is way to simplify this. Why do I need login command even? What any command that needs Auth do automatic renewal if needed etc?

Reply:
The flow would be something like, 

We need to perform operations for `tenant1` so we run `obsctl login <tenant1 auth flags>` which doesn't do any HTTP operation but saves the tenant config to a local config file and sets the current tenant to `tenant1`. Any subsequent `obsctl` operations (which will have HTTP calls) would use config for `tenant1`.

Next, we might want to switch to `tenant2` for other operations, so would need to login as `tenant2` by running `obsctl login <tenant2 auth flags>`. Any subsequent operations would then use `tenant2` config as it was set to the current tenant.

But then once we want to switch back to `tenant1` we would run `obsctl switch tenant1`. Any subsequent operation after this would again use `tenant1` configs as current tenant is again set to `tenant1`. 

obsctl would aim to follow similar semantics to kubectl [context](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts). 🙂

As for token sessions, that is something that needs to be better clarified here. Will discuss this with @onprem!

Reply:
My feeling is tenants/observatorium are more analogous to projects/namespaces in a kube cluster as opposed to completely distinct clusters. 

My understanding is switching tenants will always have to be preceded with a login command, so why not merge the two and make it so that a `switch` does a config OIDC config check/token handshake for that particular tenant? It would render the login/logout commands moot and simplify the CLI.

Reply:
I see that tenants are similar to namespaces, but a major difference is that each tenant has different creds, while in K8s your credentials are cluster level. I spent some time thinking about how this whole login flow should look like, and I think it will be even trickier that K8s. Because we have multiple instances of Observatorium API, with multiple tenants, and these tenants can have multiple users.


> What any command that needs Auth do automatic renewal if needed etc?

Since we are using OAuth2/OIDC, we can get refresh tokens from SSO provider and implement automatic token refresh flow. I think we should do that.